src/qt/rpcconsole.cpp

```diff
@@ -958,6 +959,18 @@ void RPCConsole::updateNetworkState()
     }
 
     ui->numberOfConnections->setText(connections);
+
+    // localAddressess
+    std::string localAddresses="";
+    LOCK(g_maplocalhost_mutex);
+    for (const std::pair<const CNetAddr, LocalServiceInfo> &item : mapLocalHost)
+        localAddresses += item.first.ToString() + ":" + std::to_string(item.second.nPort) + ", ";
+    if (localAddresses.length() > 2)
+        localAddresses[localAddresses.length()-2] = 0;
+    else
+        localAddresses = "None";
```

Comment:
This is a review on an earlier state of the PR, the following would be the appropriate diff to express your idea:

``` diff
diff --git a/src/qt/rpcconsole.cpp b/src/qt/rpcconsole.cpp
index 386b9373c..a249d7895 100644
--- a/src/qt/rpcconsole.cpp
+++ b/src/qt/rpcconsole.cpp
@@ -964,13 +964,12 @@ void RPCConsole::updateNetworkState()
     // localAddressess
     QString local_addresses;
     LOCK(g_maplocalhost_mutex);
-    for (const std::pair<const CNetAddr, LocalServiceInfo> &item : mapLocalHost)
-        local_addresses += QString::fromStdString(item.first.ToString()) + ":" + QString::number(item.second.nPort) + ", ";
-    if (local_addresses.length() > 2)
-        local_addresses[local_addresses.length()-2] = 0;
-    else
-        // Signals to the user that they do not have any local addresses.
-        local_addresses = tr("None");
+    for (auto it = mapLocalHost.begin(); it != mapLocalHost.end();) {
+        local_addresses += QString::fromStdString(it->first.ToString()) + ":" + QString::number(it->second.nPort) + ", ";
+        if (++it != mapLocalHost.end()) local_addresses += ", ";
+    }
+    //: Signals to the user that they do not have any local addresses.
+    if (local_addresses.isEmpty()) local_addresses = tr("None");
     ui->localAddresses->setText(local_addresses);
 }
```

Reply:
I pointed to the previous state on purpose. My main suggestion was to encapsulate the mutex and `mapLocalHost` access in the backend, which is Qt/GUI independent. So cannot use QString from there. A plain std::string that will need to be converted once it pass through the model class or directly in the widget.

Reply:
> Dear @furszy thanks for the valuable input. Applied most of them except this encapsulation part. I understand your rational and I think it's a better design. But since the number of connections are also calculated here, I did the same with Local Address.

The connections aren't calculated here (on the widget as you are doing for the external addresses). It's encapsulated inside the connections manager. The flow is the following one widget -> model -> node -> conman:

`updateNetworkState` calls `clientModel->getNumConnections` which calls the interface `node.getNodeCount`, which calls the connection manager.

Reply:
Thanks. Did the flow as you said. Now we are querying layer by layer. Please comment if it still needs improvement. I decided to pass the `std::map<CNetAddr, LocalServiceInfo>` to keep it flexible enough for different use cases; for example in this specific usecase (showing IPs in GUI), i was able to do check for "IsI2P" on CNetAddrs.